feg/gateway/services/envoy_controller/control_plane/control_plane.go

```diff
@@ -0,0 +1,364 @@
+	var grpcOptions []grpc.ServerOption
+	grpcOptions = append(grpcOptions, grpc.MaxConcurrentStreams(grpcMaxConcurrentStreams))
+	grpcServer := grpc.NewServer(grpcOptions...)
+
+	lis, err := net.Listen("tcp", fmt.Sprintf(":%d", port))
+	if err != nil {
+		glog.Fatalf("failed to listen %s", err)
+	}
+
+	// register services
+	discovery.RegisterAggregatedDiscoveryServiceServer(grpcServer, server)
+	v2.RegisterEndpointDiscoveryServiceServer(grpcServer, server)
+	v2.RegisterClusterDiscoveryServiceServer(grpcServer, server)
+	v2.RegisterRouteDiscoveryServiceServer(grpcServer, server)
+	v2.RegisterListenerDiscoveryServiceServer(grpcServer, server)
+
+	glog.Infof("Management server listening on port %d", port)
+	go func() {
+		if err = grpcServer.Serve(lis); err != nil {
+			glog.Error(err)
+		}
+	}()
+	<-ctx.Done()
+
+	grpcServer.GracefulStop()
+}
+
+// RunManagementGateway starts an HTTP gateway to an xDS server.
+func RunManagementGateway(ctx context.Context, srv xds.Server, port uint) {
+	glog.Infof("Gateway listening HTTP/1.1 on port %d", port)
+	server := &http.Server{Addr: fmt.Sprintf(":%d", port), Handler: &xds.HTTPGateway{Server: srv}}
+	go func() {
+		if err := server.ListenAndServe(); err != nil {
+			glog.Error(err)
+		}
+	}()
+}
+
+func GetControllerClient() *ControllerClient {
+	cli := ControllerClient{}
+	ctx := context.Background()
+
+	glog.Infof("Starting Envoy control plane")
+
+	signal := make(chan struct{})
+	cb := &callbacks{
+		signal:   signal,
+		fetches:  0,
+		requests: 0,
+	}
+	cli.config = cache.NewSnapshotCache(mode == Ads, Hasher{}, nil)
+
+	srv := xds.NewServer(ctx, cli.config, cb)
+
+	// start the xDS server
+	go RunManagementServer(ctx, srv, port)
+	go RunManagementGateway(ctx, srv, gatewayPort)
+
+	<-signal
+
+	cb.Report()
+
+	return &cli
+}
+
+func getHttpConnectionManager(routeConfigName string, virtualHosts []*v2route.VirtualHost) *hcm.HttpConnectionManager {
+	return &hcm.HttpConnectionManager{
+		CodecType:        hcm.HttpConnectionManager_AUTO,
+		StatPrefix:       "ingress_http",
+		UseRemoteAddress: &wrappers.BoolValue{Value: true},
+		CommonHttpProtocolOptions: &core.HttpProtocolOptions{
+			IdleTimeout: ptypes.DurationProto(idleTimeout),
+			// TODO figure out why this doesn't work properly
+			//HeadersWithUnderscoresAction: core.HttpProtocolOptions_REJECT_REQUEST,
+		},
+		Http2ProtocolOptions: &core.Http2ProtocolOptions{
+			MaxConcurrentStreams:        &wrappers.UInt32Value{Value: maxConcurrentStreams},
+			InitialStreamWindowSize:     &wrappers.UInt32Value{Value: initialStreamWindowSize},
+			InitialConnectionWindowSize: &wrappers.UInt32Value{Value: initialConnectionWindowSize},
+		},
+		RouteSpecifier: &hcm.HttpConnectionManager_RouteConfig{
+			RouteConfig: &v2.RouteConfiguration{
+				Name:         routeConfigName,
+				VirtualHosts: virtualHosts,
+			},
+		},
+		HttpFilters: []*hcm.HttpFilter{{
+			Name: wellknown.Router,
+		}},
+	}
+}
+
```

Comment:
noramlly we don't nest structs like this (normally only on testing).

Put it nested elements is faster when you do it, but not easy when you have to re-read it after few month

```
useRemoteAddress := &wrappers.BoolValue{Value: true},
commonHttpProtolOptions := &core.HttpProtocolOptions{
			IdleTimeout: ptypes.DurationProto(idleTimeout),
			// TODO figure out why this doesn't work properly
			//HeadersWithUnderscoresAction: core.HttpProtocolOptions_REJECT_REQUEST,
		}

http2ProtocolOptions := &core.Http2ProtocolOptions{
			MaxConcurrentStreams:        &wrappers.UInt32Value{Value: maxConcurrentStreams},
			InitialStreamWindowSize:     &wrappers.UInt32Value{Value: initialStreamWindowSize},
			InitialConnectionWindowSize: &wrappers.UInt32Value{Value: initialConnectionWindowSize},
		}

routeSpecifier := &hcm.HttpConnectionManager_RouteConfig{
			RouteConfig: &v2.RouteConfiguration{
				Name:         routeConfigName,
				VirtualHosts: virtualHosts,
			},
		},

httpFilters := []*hcm.HttpFilter{{
			Name: wellknown.Router,
		}},


return &hcm.HttpConnectionManager{
		CodecType:        hcm.HttpConnectionManager_AUTO,
		StatPrefix:       "ingress_http",
		UseRemoteAddress:  useRemoteAddress,
		CommonHttpProtocolOptions: commonHttpProtolOptions,
		Http2ProtocolOptions: http2ProtocolOptions,
		RouteSpecifier: routeSpecifier,
		HttpFilters: httpFilters,
	}